### PR TITLE
Fix/ingress class

### PR DESCRIPTION
**Pull request by ycliuhw**
*The ingress created by juju expose now will try to use the IngressClassName as a fallback if there is a default IngressClass resource found*

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

## QA steps

1. use the ingressclass that provided;
```console
$ juju add-model t1 microk8s

$ juju deploy cs:~juju/mariadb-k8s-3

$ juju deploy cs:~juju/mediawiki-k8s-4 --config juju-external-hostname=172.16.104.2.xip.io --config=kubernetes-ingress-class=foo

$ juju relate mediawiki-k8s:db mariadb-k8s:server

$ juju expose mediawiki-k8s

$ mkubectl get ingressclass
NAME     CONTROLLER             PARAMETERS   AGE
public   k8s.io/ingress-nginx   <none>       41h

$ mkubectl get ing mediawiki-k8s -n t1 -o json | jq '{annotations: .metadata.annotations, ingressClassName: .spec.ingressClassName}'

{
  "annotations": {
    "ingress.kubernetes.io/rewrite-target": "",
    "ingress.kubernetes.io/ssl-passthrough": "false",
    "ingress.kubernetes.io/ssl-redirect": "false",
    "kubernetes.io/ingress.allow-http": "false",
    "kubernetes.io/ingress.class": "foo"
  },
  "ingressClassName": null
}

```

2. use `ingressclass` resource as a fallback;
```console
$ juju add-model t1 microk8s

$ juju deploy cs:~juju/mariadb-k8s-3

$ juju deploy cs:~juju/mediawiki-k8s-4 --config juju-external-hostname=172.16.104.2.xip.io

$ juju relate mediawiki-k8s:db mariadb-k8s:server

$ juju expose mediawiki-k8s

$ curl -s -o /dev/null -w "%{http_code}" 172.16.104.2.xip.io

$ mkubectl get ingressclass
NAME     CONTROLLER             PARAMETERS   AGE
public   k8s.io/ingress-nginx   <none>       41h

$ mkubectl get ing mediawiki-k8s -n t1 -o json | jq '{annotations: .metadata.annotations, ingressClassName: .spec.ingressClassName}'

{
  "annotations": {
    "ingress.kubernetes.io/rewrite-target": "",
    "ingress.kubernetes.io/ssl-passthrough": "false",
    "ingress.kubernetes.io/ssl-redirect": "false",
    "kubernetes.io/ingress.allow-http": "false"
  },
  "ingressClassName": "public"
}

```

3. use `nginx` as a fallback;
```console
$ juju add-model t1 microk8s

$ juju deploy cs:~juju/mariadb-k8s-3

$ juju deploy cs:~juju/mediawiki-k8s-4 --config juju-external-hostname=172.16.104.2.xip.io

$ juju relate mediawiki-k8s:db mariadb-k8s:server

$ juju expose mediawiki-k8s

$ mkubectl get ingressclass
NAME     CONTROLLER             PARAMETERS   AGE
No resources found

$ mkubectl get ing mediawiki-k8s -n t1 -o json | jq '{annotations: .metadata.annotations, ingressClassName: .spec.ingressClassName}'

{
  "annotations": {
    "ingress.kubernetes.io/rewrite-target": "",
    "ingress.kubernetes.io/ssl-passthrough": "false",
    "ingress.kubernetes.io/ssl-redirect": "false",
    "kubernetes.io/ingress.allow-http": "false",
    "kubernetes.io/ingress.class": "nginx"
  },
  "ingressClassName": null
}

```

## Documentation changes

No

## Bug reference

No

### CODE REVIEW

**Comment by ycliuhw** [comment]
[Current Ingress implementation is broken on microk8s 1.20/stable because it introduces `IngressClass`](https://trello.com/c/KGfRMq9e/1338-current-ingress-implementation-is-broken-on-microk8s-120-stable-because-it-introduces-ingressclass)

**Comment by ycliuhw** [comment]
!!build!!

File: caas/kubernetes/provider/ingress.go [hunk 1]
```diff
@@ -131,3 +132,17 @@ func (k *kubernetesClient) deleteIngressResources(appName string) error {
...
+func (k *kubernetesClient) listIngressClasses(labels map[string]string) ([]networkingv1.IngressClass, error) {
+	listOps := v1.ListOptions{
+		LabelSelector: utils.LabelsToSelector(labels).String(),
+	}
+	ingCList, err := k.client().NetworkingV1().IngressClasses().List(context.TODO(), listOps)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(ingCList.Items) == 0 {
+		return nil, errors.NotFoundf("ingress with labels %v", labels)
```
Comment (wallyworld) [inline caas/kubernetes/provider/ingress.go:145]: "ingress class with labels %v"
?

**Review by wallyworld — APPROVED** [review]

File: caas/kubernetes/provider/ingress.go (hunk 1 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/ingress.go:145]: ops

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**